File: examples/15_algorithms/15e_lis/lis.py
"""Longest increasing subsequence: https://www.acmicpc.net/problem/12015

Input:  N on the first line, then N integers.
Output: the length of the longest strictly increasing subsequence.
"""

import array


def longest_increasing(values, tails):
    length = 0
    for i in range(len(values)):
        value = values[i]
        low = 0
        high = length
        while low < high:
            middle = (low + high) // 2
            if tails[middle] < value:
                low = middle + 1
            else:
                high = middle
        tails[low] = value
        if low == length:
            length += 1
    return length


def read_values():
    """The judge's input, or a generated stand-in when nothing is piped in."""
    try:
        count = int(input())
    except EOFError:
        return array.array("q", [(i * 7919 + (i % 977) * 104729) % 1000003 for i in range(1000000)])
    values = array.array("q", [0] * count)
    for i in range(count):
        values[i] = int(input())
    return values


def main():
    values = read_values()
    tails = array.array("q", [0] * len(values))
    print(longest_increasing(values, tails))


main()

File: examples/15_algorithms/15e_lis/lis.c
/* The same problem as lis.ppy, reading the judge's input with scanf. */
#include <stdio.h>
#include <stdlib.h>

static long long longest_increasing(const long long *values, long long size, long long *tails) {
    long long length = 0;
    for (long long i = 0; i < size; i++) {
        long long value = values[i];
        long long low = 0;
        long long high = length;
        while (low < high) {
            long long middle = (low + high) / 2;
            if (tails[middle] < value) {
                low = middle + 1;
            } else {
                high = middle;
            }
        }
        tails[low] = value;
        if (low == length) {
            length += 1;
        }
    }
    return length;
}

int main(void) {

    long long size = 0;
    long long *values = NULL;
    if (scanf("%lld", &size) == 1) {
        values = malloc((size_t)size * sizeof(long long));
        for (long long i = 0; i < size; i++) {
            if (scanf("%lld", &values[i]) != 1) {
                return 1;
            }
        }
    } else {
        /* Nothing piped in: the same stand-in the .ppy generates. */
        size = 1000000;
        values = malloc((size_t)size * sizeof(long long));
        for (long long i = 0; i < size; i++) {
            values[i] = (i * 7919 + (i % 977) * 104729) % 1000003;
        }
    }

    long long *tails = calloc((size_t)size, sizeof(long long));
    long long answer = longest_increasing(values, size, tails);

    printf("%lld\n", answer);
    return 0;
}
